### DISCUSSION TITLE
How to calculate pollutants concentration from swmm model

### DISCUSSION THREAD

**Discussion by MohammadTahirGhori** [pyswmm/pyswmm#316] [Q&A]
when I run this to get concentration of TSS, TN, and TP as I assigned in swmm file the code says an error. I tried both for nodes and catchments. 
errors:
```
(AttributeError: function 'swmm_getNodePollut' not found)
for catchment (ValueError: NULL pointer access)
```
I am doing an optimization that I need pollutants concentration reduction after LID implementation then I will multiply the reduction of concentration by cost of treatment of TSS, TN, and TP to make it monetary value and use for optimization process with cost of NBS measures. 
Here are two of codes for nodes and catchments. 
I do not know if this is the right way or not?

```python
from pyswmm import Simulation, Subcatchments, Links, Nodes, SystemStats
with Simulation('D:/Thesis/MSc_Yang/strategy1/Basefile.inp') as sim:
    sim.step_advance(900)
    J1 = Nodes(sim)["a0-n18"]
    for step in sim:
        print(J1.pollut_quality)
```
```python
from pyswmm import Simulation, Subcatchments, Links, Nodes, SystemStats
with Simulation('D:/Thesis/MSc_Yang/strategy1/Basefile.inp') as sim:
    sim.step_advance(900)
    s1 = Subcatchments(sim)["A1"]
    for step in sim:
        print(s1.pollut_quality)
```

**Comment by bemcdonnell** [comment]
@jennwuu or @abhiramm7, routing this your way

**Comment by bemcdonnell** [comment]
@MohammadTahirGhori Can you please share the PySWMM version number, Python version, and platform?

**Reply by MohammadTahirGhori** [reply]
Hi 
Python version is 3.9
pyswmm version: actually I started importing it and using since 15 days and I followed the last version. I am not sure which version is it.
I am coding in python 3.9 spyder which I downloaded from anaconda.

**Comment by jennwuu** [comment]
@MohammadTahirGhori Is it possible to share the SWMM model input as well? Or a small subset of the model you are using.

**Reply by abhiramm7** [reply]
Thanks @MohammadTahirGhori , I'll take a look at it and get back to you :)

**Comment by abhiramm7** [comment]
@MohammadTahirGhori, I think this might be a version issue.

 I can run your code with no issues. You can use the attached script to check your pyswmm version. 

```python
import pyswmm
print(f"pyswmm version: {pyswmm.__version__}")
```
You should be able to run your code in pyswmm >= 1.0.1.  If you are using pyswmm < 1.0.1, you should consider upgrading. 

Hope this helps, do let me know if you have any questions  :)

**Reply by MohammadTahirGhori** [reply]
Thanks. 
Checked my pyswmm version is 
0.6.2

So you are able to get pollutants concentration?

**Reply by MohammadTahirGhori** [reply]
I have the pollutants introduced in swmm file with wash off concentrations for different land uses assigned for sub catchments. 
I need to have the pollutants concentration then I need to change it to Kg. if I have the (kg) of TSS, TN, and TP then by multiplying by unit cost of treatment of those pollutants I can have a cost. Then I will use the cost of treatment to put in optimization process with NSGA 2 algorithm with cost of investment of LID measures as optimization objectives. 

Can you guide me how to write a code on this process?

**Reply by abhiramm7** [reply]
You should upgrade your pyswmm to 1.0.1. Then you will be able to query pollutants. The script you have above will work. 
You can upgrade using
```python
pip install -U pyswmm
```

Once you have the pollutant query working, you can store the concentrations in an array and convert them into mass. Then you should be able to do what you have outlined above. 

Let me know if you have any specific questions.

**Reply by MohammadTahirGhori** [reply]
Thats great. 
Thanks a lot. 
If I have any question I will post here. 
Thanks again.

**Reply by MohammadTahirGhori** [reply]
@abhiramm7 
Hi 
I have updated the pyswmm but now it shows this error. 
None of the pyswmm functions work.

ImportError: DLL load failed while importing _solver: The specified procedure could not be found.

**Comment by jennwuu** [comment]
@MohammadTahirGhori Can you confirm your swmm-toolkit version as well? pyswmm dependents on swmm-toolkit

**Reply by MohammadTahirGhori** [reply]
@jennwuu 
How to do that?

**Reply by jennwuu** [reply]
`pip freeze` in command prompt and look for swmm-toolkit

**Reply by MohammadTahirGhori** [reply]
I found this 
swmm-toolkit==0.8.2

**Reply by MohammadTahirGhori** [reply]
And pyswmm version shows 

pyswmm==1.1.1

**Reply by jennwuu** [reply]
This is an error due to conda installation. The bug was fixed by @karosc, but not bug fix is not released yet in swmm-toolkit. In the meantime, can you use regular Python 3.9?

**Reply by MohammadTahirGhori** [reply]
@jennwuu 
I should use conda actually. 
How about using lower version of pyswmm? to solve the problem.

**Reply by jennwuu** [reply]
If you need to use conda, then I recommend add conda (e.g. C:\Anaconda3\bin; C:\Anaconda3\Library\bin) to your system path manually if you're using windows.

**Reply by michaeltryby** [reply]
@MohammadTahirGhori is your conda environment getting activated properly? It needs to be activated for the dll search path to get setup properly. 

Try using conda from the command prompt activate it and then at the Python prompt try to import pyswmm. What happens?

**Reply by MohammadTahirGhori** [reply]
I got this when typing python in window command.

C:\WINDOWS\system32>python
Python 3.9.7 (default, Sep 16 2021, 16:59:28) [MSC v.1916 64 bit (AMD64)] :: Anaconda, Inc. on win32

Warning:
This Python interpreter is in a conda environment, but the environment has
not been activated.  Libraries may fail to load.  To activate this environment
please see https://conda.io/activation

Type "help", "copyright", "credits" or "license" for more information.
>>>

**Reply by michaeltryby** [reply]
You need to activate the environment and then install pyswmm.

**Reply by MohammadTahirGhori** [reply]
Okay, thanks

**Reply by MohammadTahirGhori** [reply]
I followed the link regarding activating the environment like follow:
conda activate C:\Anaconda
I tried this in anaconda and windows command but it says could not find the conda environment. 
How can I activate the environment?

**Reply by jennwuu** [reply]
Try following this guide when using conda https://docs.conda.io/projects/conda/en/latest/user-guide/getting-started.html.

**Comment by MohammadTahirGhori** [comment]
![env](https://user-images.githubusercontent.com/95773533/153514901-3ac3aeaa-7f19-427b-bfa2-b863798c3838.JPG)

**Reply by michaeltryby** [reply]
Read the instructions on how to activate your conda environment.
